hsmd/hsm.c

```diff
@@ -569,6 +570,51 @@ static void sign_withdrawal_tx(struct daemon_conn *master, const u8 *msg)
 	tal_free(tmpctx);
 }
 
+/**
+ * sign_invoice - Sign an invoice with our key.
+ */
+static void sign_invoice(struct daemon_conn *master, const u8 *msg)
+{
+	const tal_t *tmpctx = tal_tmpctx(master);
+	u5 *u5bytes;
+	u8 *hrpu8;
+	char *hrp;
+	struct sha256 sha;
+        secp256k1_ecdsa_recoverable_signature rsig;
+	struct hash_u5 hu5;
+	struct privkey node_pkey;
+
+	if (!fromwire_hsmctl_sign_invoice(tmpctx, msg, NULL, &u5bytes, &hrpu8)) {
+		status_trace("Failed to parse sign_invoice: %s",
+			     tal_hex(trc, msg));
+		return;
+	}
+
+	/* FIXME: Check invoice! */
+
+	hrp = tal_dup_arr(tmpctx, char, (char *)hrpu8, tal_len(hrpu8), 1);
+	hrp[tal_len(hrpu8)] = '\0';
+
+	hash_u5_init(&hu5, hrp);
+	hash_u5(&hu5, u5bytes, tal_len(u5bytes));
+	hash_u5_done(&hu5, &sha);
+
+	node_key(&node_pkey, NULL);
+        if (!secp256k1_ecdsa_sign_recoverable(secp256k1_ctx, &rsig,
+                                              (const u8 *)&sha,
+                                              node_pkey.secret.data,
+                                              NULL, NULL)) {
+		/* FIXME: Now master will freeze... */
+		status_trace("Failed to sign invoice: %s",
+			     tal_hex(trc, msg));
+		return;
```

Comment:
Maybe adding an error field that indicates failure into the reply and return that is sufficient?

Reply:
Not even sure when this might happen, might be bad enough to kill the `hsm` and take the rest of the system down with it?

Reply:
Yes, it's the same as the other treatment of errors, and almost certainly means we're screwed, since master talks to us sync and we're not going to respond.  It's just that I commented this one :)

We should abort in these cases, but that's a separate patch.  I'll file a bug...